### PR TITLE
[ONNX][dynamo_export] Decomposition skips using custom operator

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117314

A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Nice! Would it make sense to have this exact API, but exposed to the user instead of having it implicitly used by the exporter? e.g. the torchbench would be changed to add this around the torch.onnx.dynamo_export call and leave the exporter as-is, without temporary stuff

```python
with torch.onnx._enabled_decomposition_skips():
   onnx_program = torch.onnx.dynamo_export()
```

The advantage is that we don't need another PR just to disable this when the bug is fixed. We can also reuse this private API for other operators as they show up without changing the exporter too
Maybe we could even tap this solution along with the custom op/decomposition table features too

File: test/test_public_bindings.py [hunk 1]
```diff
@@ -250,6 +250,7 @@ def test_modules_can_be_imported(self):
             "torch._inductor.codegen.cuda.cuda_kernel",
             "torch.onnx._internal.fx._pass",
             "torch.onnx._internal.fx.analysis",
+            "torch.onnx._internal.fx.decomposition_skip",
```
Comment (albanD) [inline test/test_public_bindings.py:253]: It would be best to make this module not crash on import.

File: torch/onnx/_internal/fx/decomposition_skip.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+)
+
+import torch
+from torch._decomp import decompositions
+
+_NEW_OP_NAMESPACE: str = "onnx_export"
+"""The namespace for the custom operator."""
+
+
+class DecompSkip:
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/decomposition_skip.py:31]:
```suggestion
class DecompSkip(abc.ABC):
```

was this abc?

**Review by justinchuby — APPROVED** [review]
This is neat!

**Comment by thiagocrepaldi** [comment]
> Hmm I actually would prefer adding a PR to disable this in the future, than needing to update every use site to enable this now and again in the future to remove it.

Are there many call sites? I was expecting a single change https://github.com/pytorch/pytorch/blob/main/benchmarks/dynamo/common.py#L1500 which would skip the targeted decompositions during torchbench, only

**Comment by BowenBao** [comment]
@thiagocrepaldi I see what you meant. I think we should rather apply this by default on exporter. Ideally there shouldn't be difference in the end model people get, between this method and the eventual fix. So this is implementation details.

File: torch/onnx/_internal/fx/decomposition_skip.py (hunk 2 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/decomposition_skip.py:31]: I'll likely add this in a follow up now that all tests are green.. Depending on if we have permission to merge.

File: test/test_public_bindings.py (hunk 1 above)
Comment (BowenBao) [reply test/test_public_bindings.py:253]: @albanD could you approve? Please let me know if this is a concern.

**Review by malfet — APPROVED** [review]
